core/state/snapshot/snap_iterator.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
+package snapshot
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+// snapIter is a wrapper of underlying database iterator. It extends
+// the basic iteration interface by adding Discard which can cache
+// the element internally where the iterator is currently located.
+type snapIter struct {
```

Comment:
This name is a bit strange, as it doesn't have anything to do with snapshots or this package really (apart from needing the extra flexibility). I guess it makes sense to keep it local to this package being super specific, but could we reame it to `holdableIterator`? I understand the name is a bit wonky, but at least it lends some info towards what it does. Also let's not shorter `iterator` to `iter` in the type name. The abbreviation is often used as variable and all other types explicitly name it `XyzIterator`.

Reply:
Changed.